File: src/main/javascript/crypto/src/libs/AbstractSignature.ts
import {Signature} from "./Signature";
import {KeyPair} from "./KeyPair";
import {hexStringToArray, uint8tohex} from "./utils";

export abstract class AbstractSignature implements Signature {
    private type: string;
    private rawSignature: string;

    public keyMessageType(keys: KeyPair, unprocessedMessage: Uint8Array, type:string) {
        this.type = type;
        this.rawSignature = this.sign(keys, unprocessedMessage);
    }

    public signatureAndType(rawSignature: string, type: string) {
        this.type = type;
        this.rawSignature = rawSignature;
    }

    protected sign(keys: KeyPair, unprocessedMessage: Uint8Array): string {
        return keys.signRawBytesWithEthereum(this.processMessage(unprocessedMessage));
    }

    public getRawSignature(): string {
        return this.rawSignature;
    }

    public getTypeOfSignature(): string {
        return this.type;
    }

    public abstract processMessage(unprocessedMsg: Uint8Array): number[];

    public verify(unprocessedMsg: Uint8Array, verificationKey: KeyPair): boolean {
        return verificationKey.verifyBytesWithEthereum(this.processMessage(unprocessedMsg), this.rawSignature);
    }

}

File: src/main/java/org/tokenscript/attestation/core/AbstractSignature.java
package org.tokenscript.attestation.core;

import org.bouncycastle.crypto.params.AsymmetricKeyParameter;

public abstract class AbstractSignature implements Signature {
  private final String type;
  private final byte[] rawSignature;

  public AbstractSignature(AsymmetricKeyParameter signingKey, byte[] unprocessedMessage, String type) {
    this.type = type;
    this.rawSignature = sign(signingKey, unprocessedMessage);
  }

  public AbstractSignature(byte[] rawSignature, String type) {
    this.type = type;
    this.rawSignature = rawSignature;
  }

  protected byte[] sign(AsymmetricKeyParameter keys, byte[] unprocessedMessage) {
    return SignatureUtility.signWithEthereum(processMessage(unprocessedMessage), keys);
  }

  @Override
  public byte[] getRawSignature() {
    return rawSignature;
  }

  @Override
  public String getTypeOfSignature() {
    return type;
  }

  /**
   * Processes any message and returns the raw bytes that are actually being signed
   * @return
   */
  abstract byte[] processMessage(byte[] unprocessedMsg);

  @Override
  public boolean verify(byte[] unprocessedMsg, AsymmetricKeyParameter verificationKey) {
    return SignatureUtility.verifyEthereumSignature(processMessage(unprocessedMsg), rawSignature, verificationKey);
  }

}
